File: bindings/nodejs/examples/kem.ts
/**
 * ML-KEM (Kyber) Key Encapsulation example
 */

import {
    initialize,
    cleanup,
    openFirstDevice,
    KemAlgorithm,
} from '../src';

async function main() {
    console.log('=== ML-KEM (Kyber) Key Encapsulation Example ===\n');

    // Initialize
    initialize();
    const device = openFirstDevice();
    const kem = device.kem;

    // Test all KEM variants
    const algorithms = [
        { algo: KemAlgorithm.MlKem512, name: 'ML-KEM-512' },
        { algo: KemAlgorithm.MlKem768, name: 'ML-KEM-768' },
        { algo: KemAlgorithm.MlKem1024, name: 'ML-KEM-1024' },
    ];

    for (const { algo, name } of algorithms) {
        console.log(`--- ${name} ---`);

        // Generate key pair
        console.log('Generating key pair...');
        const startGen = performance.now();
        const keypair = kem.generateKeypair(algo);
        const genTime = performance.now() - startGen;

        console.log(`  Public key: ${keypair.publicKey.length} bytes`);
        console.log(`  Secret key: ${keypair.secretKey.length} bytes`);
        console.log(`  Time: ${genTime.toFixed(2)}ms`);

        // Encapsulate (sender side)
        console.log('Encapsulating...');
        const startEncap = performance.now();
        const { ciphertext, sharedSecret: senderSecret } = kem.encapsulate(
            keypair.publicKey,
            algo
        );
        const encapTime = performance.now() - startEncap;

        console.log(`  Ciphertext: ${ciphertext.length} bytes`);
        console.log(`  Shared secret: ${senderSecret.toString('hex').slice(0, 32)}...`);
        console.log(`  Time: ${encapTime.toFixed(2)}ms`);

        // Decapsulate (receiver side)
        console.log('Decapsulating...');
        const startDecap = performance.now();
        const receiverSecret = kem.decapsulate(keypair.secretKey, ciphertext, algo);
        const decapTime = performance.now() - startDecap;

        console.log(`  Shared secret: ${receiverSecret.toString('hex').slice(0, 32)}...`);
        console.log(`  Time: ${decapTime.toFixed(2)}ms`);

        // Verify secrets match
        const match = senderSecret.equals(receiverSecret);
        console.log(`  Secrets match: ${match ? '✓' : '✗'}`);

        if (!match) {
            console.error('ERROR: Shared secrets do not match!');
            process.exit(1);
        }

        console.log();
    }

    // Convenience method example
    console.log('--- Using Convenience Methods ---');
    const keypair768 = kem.generateKeypair768();
    const { ciphertext, sharedSecret } = kem.encapsulate768(keypair768.publicKey);
    const decrypted = kem.decapsulate768(keypair768.secretKey, ciphertext);
    console.log(`Secrets match: ${sharedSecret.equals(decrypted) ? '✓' : '✗'}\n`);

    // Clean up
    device.close();
    cleanup();
    console.log('Done!');
}

main().catch(console.error);

File: bindings/rust/examples/kem.rs
//! Key Encapsulation Mechanism (KEM) example.
//!
//! This example demonstrates ML-KEM key exchange between two parties.

use quantacore::{initialize, cleanup, open_first_device, KemAlgorithm};
use quantacore::utils::to_hex;

fn main() -> quantacore::Result<()> {
    println!("QUAC 100 SDK - ML-KEM Example");
    println!("==============================\n");

    // Initialize
    initialize()?;
    let device = open_first_device()?;
    let kem = device.kem();

    // Test all algorithm variants
    for algo in [
        KemAlgorithm::MlKem512,
        KemAlgorithm::MlKem768,
        KemAlgorithm::MlKem1024,
    ] {
        println!("Testing {}:", algo);
        println!("  Public key size: {} bytes", algo.public_key_size());
        println!("  Secret key size: {} bytes", algo.secret_key_size());
        println!("  Ciphertext size: {} bytes", algo.ciphertext_size());
        println!();

        // Alice generates a key pair
        println!("  Alice: Generating key pair...");
        let alice_keypair = kem.generate_keypair(algo)?;
        println!("    Public key: {}...", &to_hex(alice_keypair.public_key())[..32]);

        // Bob encapsulates using Alice's public key
        println!("  Bob: Encapsulating...");
        let (ciphertext, bob_secret) = kem.encapsulate(
            alice_keypair.public_key(),
            algo,
        )?;
        println!("    Ciphertext: {}...", &to_hex(&ciphertext)[..32]);
        println!("    Shared secret: {}", to_hex(&bob_secret));

        // Alice decapsulates using her secret key
        println!("  Alice: Decapsulating...");
        let alice_secret = kem.decapsulate(
            alice_keypair.secret_key(),
            &ciphertext,
            algo,
        )?;
        println!("    Shared secret: {}", to_hex(&alice_secret));

        // Verify they match
        assert_eq!(alice_secret, bob_secret);
        println!("  ✓ Shared secrets match!\n");
    }

    // Using convenience methods
    println!("Using convenience methods (ML-KEM-768):");
    let keypair = kem.generate_keypair_768()?;
    let (ct, sender_ss) = kem.encapsulate_768(keypair.public_key())?;
    let receiver_ss = kem.decapsulate_768(keypair.secret_key(), &ct)?;
    assert_eq!(sender_ss, receiver_ss);
    println!("  ✓ Key exchange successful!\n");

    // Cleanup
    drop(device);
    cleanup()?;

    println!("Done.");
    Ok(())
}